### DISCUSSION TITLE
Bad result for ADC read_uv() after sleeping with lightsleep()?

### DISCUSSION THREAD

**Discussion by brianwyld** [micropython/micropython#12395] [ESP32] [answered]
On ESP32S3, I create an ADC pin, and read it using read_uv() (converting to mV in the example below).
```
>>> import machine
>>> adc = machine.ADC(machine.Pin(14, mode=machine.Pin.IN, pull=None), atten=machine.ADC.ATTN_11DB)
>>> adc.read_uv()/1000
2008.0
>>> adc.read_uv()/1000
2008.0
```
So thats fine (this is the value I expect, basically half of my battery voltage due to resistor divider)
But if I sleep a little:
```
>>> machine.lightsleep(1000)
machine.lightsleep(1000)
>>> adc.read_uv()/1000
3109.0
```
and thereafter the value is bad... A reset will return to the good value, but just re-creating the ADC pin or doing adc.init() won't.

What needs to be reinitialised/started to get the ADC to work after sleeping?

Thanks

**Comment by davefes** [comment]
Are you using the WiFi on that board?

**Reply by brianwyld** [reply]
Not currently no

**Comment by davefes** [comment]
I wasn't aware that you could set `mode`.  If that does have an effect then maybe you need to `hold` that pin during lightsleep.

```py
adc.init(hold=True)
lightsleep()
adc.int(hold=False)
```
also the line:
```py
adc = machine.ADC(machine.Pin(14, mode=machine.Pin.IN, pull=None), atten=machine.ADC.ATTN_11DB)
```
might need changing.

What happens if you leave `mode` out?

**Comment by robert-hh** [comment]
Did you measure the voltage at Pin14 when it tells 3109.0? e.g. with a multimeter

**Reply by brianwyld** [reply]
yes, the voltage is correct the first time and doesn't change between the 2 reads.

**Comment by brianwyld** [comment, marked as answer]
After some digging I found its a ESP32S3 chip bug which causes it to lose its ADC calibration register values during lightsleep().

There is a commit for esp_idf_4.4 to fix it: https://github.com/espressif/esp-idf/commit/32afe6a4985cd76e73ffa450ff070aaae4e5f93a
(essentially just adds a save/restore around the sleep)

Strangely my esp_idf_4.4 download doesnt have the fix, so I just added the save/restore functions to my esp32ext module (along with the gpio wake setup I needed to wake on a gpio on this MCU)
```

// workaround for chip bug, not in 4.4?
#define REGI2C_ANA_CALI_BYTE_NUM 8
static DRAM_ATTR uint8_t _adc_reg_val[REGI2C_ANA_CALI_BYTE_NUM];

// save ADC calibration values before sleeping (only required on ESP32S3 apparently)
STATIC mp_obj_t esp32ext_adc_cal_save (void) {
#if CONFIG_IDF_TARGET_ESP32S3
    for (int i = 0; i < REGI2C_ANA_CALI_BYTE_NUM; i++) {
        _adc_reg_val[i] = regi2c_ctrl_read_reg(I2C_SAR_ADC, I2C_SAR_ADC_HOSTID, i);
    }
#endif
    return mp_const_true;
}
STATIC MP_DEFINE_CONST_FUN_OBJ_0(esp32ext_adc_cal_save_obj, esp32ext_adc_cal_save);
// restore ADC calibration values after sleeping
STATIC mp_obj_t esp32ext_adc_cal_restore (void) {
#if CONFIG_IDF_TARGET_ESP32S3
    for (int i = 0; i < REGI2C_ANA_CALI_BYTE_NUM; i++) {
        regi2c_ctrl_write_reg(I2C_SAR_ADC, I2C_SAR_ADC_HOSTID, i, _adc_reg_val[i]);
    }
#endif
    return mp_const_true;
}
STATIC MP_DEFINE_CONST_FUN_OBJ_0(esp32ext_adc_cal_restore_obj, esp32ext_adc_cal_restore);

```
If you have a later batch of the chip, or a more recent IDF, you might not need this. In any case, its not a micropython bug.
